### PR TITLE
layers: Make many state tracker data members protected or private

### PR DESCRIPTION
**Pull request by jeremyg-lunarg**
Reduce visibility of state tracker data members, especially the state maps, which are now all private.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -3269,18 +3269,18 @@ void CoreChecks::PostCallRecordQueueSubmit2KHR(VkQueue queue, uint32_t submitCou
 }
 
 bool CoreChecks::SemaphoreWasSignaled(VkSemaphore semaphore) const {
-    for (auto &pair : queue_map_) {
-        const auto &queue_state = pair.second;
-        for (const auto &submission : queue_state->submissions) {
+    bool result = false;
+    ForEach<QUEUE_STATE>([&result, semaphore](const QUEUE_STATE &queue_state) {
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp:3273]: Wouldn't the comparable concept be AnyOf (which would, or technically is allowed to, abort on the first true).

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -4283,7 +4286,7 @@ bool CoreChecks::ValidateImageImportedHandleANDROID(const char *func_name, VkExt
 bool CoreChecks::PreCallValidateAllocateMemory(VkDevice device, const VkMemoryAllocateInfo *pAllocateInfo,
                                                const VkAllocationCallbacks *pAllocator, VkDeviceMemory *pMemory) const {
     bool skip = false;
-    if (mem_obj_map_.size() >= phys_dev_props.limits.maxMemoryAllocationCount) {
+    if (Count<DEVICE_MEMORY_STATE>() >= phys_dev_props.limits.maxMemoryAllocationCount) {
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp:4289]: "One memory state, two memory states, three memory states... ahahaha" --- wait, wrong Count ;)

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -14166,13 +14168,11 @@ bool CoreChecks::ValidateCreateSwapchain(const char *func_name, VkSwapchainCreat
     if (!instance_extensions.vk_khr_android_surface) {
         // restrict search only to queue families of VkDeviceQueueCreateInfos, not the whole physical device
         bool is_supported = false;
-        for (const auto &queue_entry : queue_map_) {
-            auto qfi = queue_entry.second->queueFamilyIndex;
-            if (surface_state->GetQueueSupport(physical_device, qfi)) {
+        ForEach<QUEUE_STATE>([this, &is_supported, surface_state](const QUEUE_STATE &queue_state) {
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp:14171]: Looks like another place for an AnyOf

File: layers/state_tracker.h [hunk 4]
```diff
@@ -329,6 +296,26 @@ class ValidationStateTracker : public ValidationObject {
...
+    size_t Count() const {
+        using Traits = AccessorTraits<State>;
+        auto map_member = Traits::Map();
+        const typename Traits::MapType& map =
+            (Traits::kInstanceScope && (this->*map_member).size() == 0) ? instance_state->*map_member : this->*map_member;
+        return map.size();
+    }
+
+    template <typename State>
+    void ForEach(std::function<void(const State& s)> fn) const {
```
Comment (jzulauf-lunarg) [inline layers/state_tracker.h:309]: You might be able to common implement ForEach and AnyOf with an "early exit" trait as a template argument.

File: layers/state_tracker.h [hunk 5]
```diff
@@ -329,6 +296,26 @@ class ValidationStateTracker : public ValidationObject {
         }
     }
 
+    template <typename State>
+    size_t Count() const {
+        using Traits = AccessorTraits<State>;
```
Comment (jzulauf-lunarg) [inline layers/state_tracker.h:301]: could these repeated map look-up lines be in a template function returning the map reference... there's getting to be quite a few copies of them.

File: layers/state_tracker.h (hunk 4 above)
Comment (jzulauf-lunarg) [reply layers/state_tracker.h:309]: ... though getting the return value right might be to much template magic... and the trade off between "DRY" and maintainable might favor... duplication, especially if the map lookup got factored out to a common function.

**Review by jzulauf-lunarg — COMMENTED** [review]
Some considerations.  Would like the loop equivalence to be maintained by going to AnyOf where needed, though not required.

**Comment by jeremyg-lunarg** [comment]
Good suggestions @jzulauf-lunarg, thank you!

File: layers/state_tracker.h [hunk 6]
```diff
@@ -264,47 +262,25 @@ class ValidationStateTracker : public ValidationObject {
         using MapType = layer_data::unordered_map<HandleType, MappedType>;
     };
 
+    template <typename State, typename Traits = AccessorTraits<State>>
```
Comment (jzulauf-lunarg) [inline layers/state_tracker.h:265]: This is way DRY'r.  +100

**Review by jzulauf-lunarg — APPROVED** [review]
LGTM

**Pull request merged**